### DISCUSSION TITLE
How to use function parameters in Modifiers?

### DISCUSSION THREAD

**Discussion by Tanmay53** [Cyfrin/foundry-full-course-cu#627] [General]
Recently, I was puzzled how to use function parameters in a modifier for that function.

Initially, I thought we could use `msg.data` to get function parameters, but `msg.data` returns ***bytes*** and I was unable to convert it to function paramters.

Then after researching more I found a very simple and quiet obvious solution for it. ✌️ 

```solidity
contract C {
  modifier restrictTargetType(address target, uint contractType) {
    require(contractTypes[target] == contractType);
    _;
  }
  function f(address target, uint contractType) restrictTargetType(target, contractType) {
    // function body
  }
}
```

**Comment by devrajsinghrawat** [comment]
This could be the suggested solution But it is worth trying to extract parameters from msg.data as in msg.data 

- 1st 4 bytes are function signature (also known as function selector) and rest are the function parameters.

**Reply by Tanmay53** [reply]
Yeah bro, do you know a way to convert the bytes to the actual data sent as function parameters?

**Reply by devrajsinghrawat** [reply]
I believe the mes.data is not in byte but in Hex .. like address is 20 bytes and it must be converted to 32 bytes by adding padding of 0.. you can take the next 32 bytes as address and cast it to address same thing you can do with contract type ..

**Reply by Tanmay53** [reply]
Bro, that bounced of my head! 🤯
